### PR TITLE
Rendering optimization and refactor

### LINKED ISSUE #3: Collect required lines instead of using them all
Right now, all the source lines are just put into an array and indexed into when necessary. This of course forces the entire list of lines, which is not the best for performance nor space. Instead, we could collect all the required lines into an `IntMap` or similar by using the line numbers and styling options from the given `Errata`.

The `makeSourceLines` function within the `renderSourceLines` function contains the logic for which lines are required. It might also be worth investigating refactoring the entire `renderSourceLines` function so that it isn't so big.

### PR DESCRIPTION
**Pull request by RiugaBachi**
:reimumug:

Sorry for the delays, kept my word in the end since I am (finally) feeling a bit better today.

e8439932fcd87cbc9ada26a410920aac4275b794 contains a perfectly good IntMap-only implementation. I'm happy to submit another pull request tomorrow for a benchmark suite if that is something you would like to see. If there's ever a need for comparison between that and the current optimized approach, I would just branch from that commit and merge benchmarks + test suite.

### CODE REVIEW

File: src/Errata/Internal/Render.hs [hunk 1]
```diff
@@ -1,6 +1,8 @@
-{-# LANGUAGE MultiWayIf        #-}
-{-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE RecordWildCards   #-}
+{-# LANGUAGE MultiWayIf          #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE ViewPatterns #-}
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: Align the `#-}`.

File: src/Errata/Internal/Render.hs [hunk 2]
```diff
@@ -36,50 +38,110 @@ renderErrors :: Source source => source -> [Errata] -> TB.Builder
...
 -- | A single pretty error from metadata and source lines.
-renderErrata :: Source source => A.Array Int source -> Errata -> TB.Builder
+renderErrata :: Source source => [source] -> Errata -> TB.Builder
 renderErrata slines (Errata {..}) = errorMessage
     where
+        -- | Header block + body blocks
+        blocks = errataBlock : errataBlocks
+
+        -- The pointers grouped by line.
+        blockPointersGrouped = I.fromListWith (<>) . map (pointerLine &&& pure) . blockPointers <$> blocks
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: I don't think the `Control.Arrow` import is justified for just this one use of `(&&&)` :P

File: src/Errata/Internal/Render.hs [hunk 3]
```diff
@@ -21,10 +23,10 @@ module Errata.Internal.Render
     , renderSourceLines
     ) where
 
-import qualified GHC.Arr as A
+import           Control.Arrow
+import           Control.Applicative
 import           Data.List
-import qualified Data.List.NonEmpty as N
-import qualified Data.Map.Strict as M
+import qualified Data.IntMap as I
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs:27]: Put this in alphabetical order please! One day I'll have a formatter I like and enforce it, but not today.

File: src/Errata/Internal/Render.hs [hunk 4]
```diff
@@ -36,50 +38,110 @@ renderErrors :: Source source => source -> [Errata] -> TB.Builder
...
-        slines = let xs = sourceToLines source in A.listArray (0, length xs - 1) xs
+        slines = sourceToLines source
         prettified = map (renderErrata slines) sortedErrata
 
 -- | A single pretty error from metadata and source lines.
-renderErrata :: Source source => A.Array Int source -> Errata -> TB.Builder
+renderErrata :: Source source => [source] -> Errata -> TB.Builder
 renderErrata slines (Errata {..}) = errorMessage
     where
+        -- | Header block + body blocks
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: `--` instead of `-- |`

File: src/Errata/Internal/Render.hs [hunk 5]
```diff
@@ -36,50 +38,110 @@ renderErrors :: Source source => source -> [Errata] -> TB.Builder
...
+        minPointers = maybe 1 id . fmap fst . I.lookupMin <$> blockPointersGrouped
+        maxPointers = maybe 0 id . fmap fst . I.lookupMax <$> blockPointersGrouped
+
+        minLine = minimum minPointers
+        maxLine = maximum maxPointers
+
+        -- Turns a list into a list of tail slices of the original list,
+        -- each element at index @i@ dropping the first @i@ elements of the original list.
+        slices [] = repeat (repeat mempty)
+        slices xs = (xs <> repeat mempty) : slices (tail xs)
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs:68]: Are the `mempty`s here just `[]`? Why the need to repeat infinitely? I'm guessing to allow out-of-bounds access? Do add a comment and a type signature.

File: src/Errata/Internal/Render.hs [hunk 6]
```diff
@@ -98,50 +160,47 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
...
-        makeSourceLines :: Line -> [Line] -> [TB.Builder]
+        makeSourceLines :: Line -> [(Line, source)] -> [TB.Builder]
 
         -- No lines left.
         makeSourceLines _ [] = []
 
         -- The next line is a line we have to decorate with pointers.
-        makeSourceLines _ (n:ns)
-            | Just p <- M.lookup n pointersGrouped = makeDecoratedLines p <> makeSourceLines 0 ns
+        makeSourceLines _ (pr@(n,_):ns)
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: Should add a space after the `,`

File: src/Errata/Internal/Render.hs [hunk 7]
```diff
@@ -98,50 +160,47 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
...
+            | Just p <- I.lookup n pointersGrouped = makeDecoratedLines p pr <> makeSourceLines 0 ns
 
         -- The next line is an extra line, within a limit (currently 2, may be configurable later).
-        makeSourceLines extra (n:ns)
+        makeSourceLines extra ((n,l):ns)
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: ^

File: src/Errata/Internal/Render.hs [hunk 8]
```diff
@@ -151,69 +210,69 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
 
                 -- There is a single extra line, so we can use that as the before-line.
                 -- No need for omission, because it came right before.
-                ([n], _) ->
+                ([(n,l)], _) ->
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: ^

File: src/Errata/Internal/Render.hs [hunk 9]
```diff
@@ -151,69 +210,69 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
...
                             | hasConnMulti       -> "  "
                             | otherwise          -> ""
-                    in (linePrefix n <> mid <> showLine [] n) : makeSourceLines 0 ns'
+                    in (linePrefix n <> mid <> showLine [] l) : makeSourceLines 0 ns'
 
                 -- There are more than one line in between, so we omit all but the last.
                 -- We use the last one as the before-line.
                 (_, _) ->
-                    let n = last es
+                    let (n,l) = last es
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: ^

File: src/Errata/Internal/Render.hs [hunk 10]
```diff
@@ -151,69 +210,69 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
...
                             | otherwise          -> ""
-                    in (omitPrefix <> mid) : (linePrefix n <> mid <> showLine [] n) : makeSourceLines 0 ns'
+                    in (omitPrefix <> mid) : (linePrefix n <> mid <> showLine [] l) : makeSourceLines 0 ns'
 
         -- Decorate a line that has pointers.
         -- The pointers we get are assumed to be all on the same line.
-        makeDecoratedLines :: N.NonEmpty Pointer -> [TB.Builder]
-        makeDecoratedLines pointers = (linePrefix line <> TB.fromText lineConnector <> sline) : decorationLines
+        makeDecoratedLines :: [Pointer] -> (Line, source) -> [TB.Builder]
+        makeDecoratedLines pointers (num,line) =
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: ^

File: src/Errata/Internal/Render.hs [hunk 11]
```diff
@@ -151,69 +210,69 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
...
-                line = pointerLine $ N.head pointers
-                sline = showLine (map pointerColumns (N.toList pointersSorted)) line
+                -- The line number we're on.
+                sline = showLine (map pointerColumns pointersSorted) line
 
                 -- The resulting decoration lines.
                 decorationLines = if
                     -- There's only one pointer, so no need for more than just an underline and label.
-                    | N.length pointersSorted' == 1                           -> [underline pointersSorted']
+                    | length pointersSorted' == 1                             -> [underline pointersSorted']
```
Comment (1Computer1) [inline src/Errata/Internal/Render.hs]: This is pretty big spacing, so I think you should just remove the horizontal alignment here.

File: src/Errata/Source.hs [hunk 12]
```diff
@@ -21,7 +21,7 @@ A class for manipulating and converting source text.
 
 For @ByteString@ source types, you should convert it to one of the built-in instances with your encoding of choice.
 -}
-class Source s where
+class Monoid s => Source s where
```
Comment (1Computer1) [inline src/Errata/Source.hs:26]: You should probably document what is expected of the `Monoid` requirement.

**Review by 1Computer1 — CHANGES_REQUESTED** [review]

**Comment by 1Computer1** [comment]
A benchmark suite would be very nice if you have the time. Would it be the adhoc one you used, or an actual suite with `criterion`?

**Comment by RiugaBachi** [comment]
A comprehensive one with `criterion`.

**Comment by 1Computer1** [comment]
That sounds great then. I'll take one last look over this in the morning and merge it then.

**Review by 1Computer1 — APPROVED** [review]

**Pull request merged**